### PR TITLE
make singleThreaded groupBy query config overridable at query time

### PR DESCRIPTION
**Pull request by himanshug**
some users found that you get better throughput by using singleThreaded processing for small groupBy queries while parallel for large groupBy queries. this allows both to be used appropriately when user knows how much data groupBy query is going to process. also, this allows rapid benchmarking for both on same cluster and data.

### CODE REVIEW

**Comment by binlijin** [comment]
👍

File: docs/content/configuration/broker.md [hunk 1]
```diff
@@ -67,7 +67,6 @@ The broker uses processing configs for nested groupBy queries. And, optionally,
 
 |Property|Description|Default|
 |--------|-----------|-------|
-|`druid.query.groupBy.singleThreaded`|Run single threaded group By queries.|false|
```
Comment (b-slim) [inline docs/content/configuration/broker.md]: i propose to keep this and add pointer or description of the new context feature. i don't think it is an overlapping with the description you have added at `query-context`.

**Comment by b-slim** [comment]
minor comments, but overall :+1:

File: processing/src/main/java/io/druid/query/QueryRunnerFactory.java [hunk 2]
```diff
@@ -57,7 +57,7 @@
    * @param queryRunners Individual QueryRunner objects that produce some results
    * @return a QueryRunner that, when asked, will use the ExecutorService to runt he base QueryRunners
    */
-  public QueryRunner<T> mergeRunners(ExecutorService queryExecutor, Iterable<QueryRunner<T>> queryRunners);
+  public QueryRunner<T> mergeRunners(ExecutorService queryExecutor, Iterable<QueryRunner<T>> queryRunners, QueryType query);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/QueryRunnerFactory.java]: Here we're introducing the concept of context for the merge task. The query runner factory hasn't had context itself prior to this. Context is often used within the runner itself. I don't have a good solution off the top of my head yet, but are we sure we want to introduce context at this level?

File: docs/content/configuration/broker.md (hunk 1 above)
Comment (xvrl) [inline docs/content/configuration/broker.md]: if we want to be backwards compatible we should have this configuration option define the default mode, but have it overridable at query time.

Comment (himanshug) [inline docs/content/configuration/broker.md]: i have just removed it from the broker docs because it does not seem to be doing anything at the broker. i think, it only applies to historicals.
default is still totally overridableby setting druid.query.groupBy.singleThreaded property at historicals and historical doc already has it.

File: processing/src/main/java/io/druid/query/QueryRunnerFactory.java (hunk 2 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/query/QueryRunnerFactory.java]: yeah, thought about that but can't find any alternative and this option looked ok to me.

**Comment by himanshug** [comment]
@xvrl , @drcrallen  ?

File: processing/src/main/java/io/druid/query/QueryRunnerFactory.java (hunk 2 above)
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/QueryRunnerFactory.java]: one possible alternative could be to move the single threaded vs multi threaded logic from QueryRunnerFactory to GroupByParallelQueryRunner, which can look at the context and choose whether to do single/multi threaded run depending on the context.

Comment (himanshug) [inline processing/src/main/java/io/druid/query/QueryRunnerFactory.java]: that is a good idea, however i still don't think of exposing query as a param in mergeRunners(..) as a problem.
will change only if you guys have strong opinion here.

Comment (himanshug) [inline processing/src/main/java/io/druid/query/QueryRunnerFactory.java]: @drcrallen @nishantmonu51 updated impl to not change mergeRunner(..) interface.

**Comment by nishantmonu51** [comment]
👍 , LGTM

File: processing/src/main/java/io/druid/query/GroupByParallelQueryRunner.java [hunk 3]
```diff
@@ -56,6 +56,8 @@
 
 public class GroupByParallelQueryRunner<T> implements QueryRunner<T>
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/GroupByParallelQueryRunner.java]: this class can do both single and parallel running, perhaps we should rename it to simply `GroupByQueryRunner`

Comment (gianm) [inline processing/src/main/java/io/druid/query/GroupByParallelQueryRunner.java]: nevermind, that already exists and is a different thing. `GroupByMergedQueryRunner`?

**Comment by gianm** [comment]
👍 other than possibly changing the name of the query runner.

**Comment by himanshug** [comment]
@gianm renamed

**Comment by gianm** [comment]
👍

File: processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java [hunk 4]
```diff
@@ -79,7 +81,108 @@ public GroupByParallelQueryRunner(
...
+            {
+              @Override
+              public QueryRunner<T> apply(final QueryRunner<T> input)
+              {
+                return new QueryRunner<T>()
+                {
+                  @Override
+                  public Sequence<T> run(final Query<T> query, final Map<String, Object> responseContext)
+                  {
+                    final Pair<IncrementalIndex, Accumulator<IncrementalIndex, T>> indexAccumulatorPair = GroupByQueryHelper
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: there's a lot of cut-and-paste between the two run methods, is there any way we can refactor that @himanshug ?

Comment (himanshug) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: @xvrl they are quite similar but just a bit different, i just moved the code around but now that I'm looking closely at the code, i think it can be refactored and improved a bit ... it seems in singleThreaded case one index is being created per runner from input set of runners while parallel one creates only one index for all of them... same should be doable for single threaded case also.
I'm making the changes, but they look a bit intrusive and unrelated to this PR. So, I propose that I will submit another PR soon with refactoring and possible improvement.

let us get this one merged as I have some people waiting for this feature to get delivered.

Comment (xvrl) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: @himanshug there's redundancy in the getting some of the query flags and creating accumulators, not sure if that can be made better. I'm only harping on this because you were so adamant about similar redundancies in #2848 even though that code already existed before, just like in your case :)

Comment (himanshug) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: @xvrl change I'm going to make is going to be a bit more than refactoring. The reason for separate PR proposal was because I wanted to remove creation of multiple indexes in single threaded case like is done for the parallel runner which changes the implementation of single threaded case really and not just moving code around for refactoring.

That said, I can see your point and will update this PR itself.

Comment (gianm) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: IMO if refactoring is not necessary to fix a bug / add a feature, I'd rather see separate PRs for the functional change and for the refactoring. It makes things easier to review. But I don't think we have an official policy on this.

Especially in this case where it sounds like the refactoring might also involve another, unrelated functional change.

Comment (himanshug) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: @xvrl @gianm code is not much , its just that i have changed the implementation of single threaded processing now along with refactoring, check the latest commit. now both are processed in same way.

Also, I don't want to start a debate here, it is ok for having disagreements on a proposal :)

Comment (xvrl) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: @himanshug if you think it's better to do a bigger refactoring then let's do that separately, no objections on that front.

Comment (himanshug) [inline processing/src/main/java/io/druid/query/GroupByMergedQueryRunner.java]: @xvrl i had done that in this PR so its taken care of.
